ext/com/concurrent_ruby/ext/JavaSemaphoreLibrary.java

```diff
@@ -0,0 +1,144 @@
+package com.concurrent_ruby.ext;
+
+import java.io.IOException;
+import java.util.concurrent.Semaphore;
+import org.jruby.Ruby;
+import org.jruby.RubyBoolean;
+import org.jruby.RubyClass;
+import org.jruby.RubyFixnum;
+import org.jruby.RubyModule;
+import org.jruby.RubyNumeric;
+import org.jruby.RubyObject;
+import org.jruby.anno.JRubyClass;
+import org.jruby.anno.JRubyMethod;
+import org.jruby.runtime.ObjectAllocator;
+import org.jruby.runtime.ThreadContext;
+import org.jruby.runtime.builtin.IRubyObject;
+
+public class JavaSemaphoreLibrary {
+
+    public void load(Ruby runtime, boolean wrap) throws IOException {
+        RubyModule concurrentMod = runtime.defineModule("Concurrent");
+        RubyClass atomicCls = concurrentMod.defineClassUnder("JavaSemaphore", runtime.getObject(), JRUBYREFERENCE_ALLOCATOR);
+
+        atomicCls.defineAnnotatedMethods(JavaSemaphore.class);
+
+    }
+
+    private static final ObjectAllocator JRUBYREFERENCE_ALLOCATOR = new ObjectAllocator() {
+        public IRubyObject allocate(Ruby runtime, RubyClass klazz) {
+            return new JavaSemaphore(runtime, klazz);
+        }
+    };
+
+    @JRubyClass(name = "JavaSemaphore", parent = "Object")
+    public static class JavaSemaphore extends RubyObject {
+
+        private JRubySemaphore semaphore;
+
+        public JavaSemaphore(Ruby runtime, RubyClass metaClass) {
+            super(runtime, metaClass);
+        }
+
+        @JRubyMethod
+        public IRubyObject initialize(ThreadContext context, IRubyObject value) {
+            this.semaphore = new JRubySemaphore(rubyFixnumToInt(value, "count"));
+            return context.nil;
+        }
+
+        @JRubyMethod
+        public IRubyObject acquire(ThreadContext context, IRubyObject value) throws InterruptedException {
+            this.semaphore.acquire(rubyFixnumToInt(value, "permits"));
+            return context.nil;
+        }
+
+        @JRubyMethod(name = "available_permits")
+        public IRubyObject availablePermits(ThreadContext context) {
+            return new RubyFixnum(getRuntime(), this.semaphore.availablePermits());
+        }
+
+        @JRubyMethod(name = "drain_permits")
+        public IRubyObject drainPermits(ThreadContext context) {
+            return new RubyFixnum(getRuntime(), this.semaphore.drainPermits());
```

Comment:
I've missed one more thing. JRuby extensions should try to use interface provided by `Ruby` object (result of `getRuntime()`) it may do additional steps like caching etc. see https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/RubyFixnum.java#L207-L212. So it's better to use the `newBoolean` and `newFixnum` methods on `Ruby` object.